### PR TITLE
Tumor FROC Evaluation

### LINKED ISSUE #1872: Add Pathology based FROC metric calculation class

### PR DESCRIPTION
**Pull request by behxyz**
Fixes #1872 

### Description
It a reimplementation of #1873 that
- Uses WSIReader to support both OpenSlide and CuCIM
- Updates class, methods, parameters, and variables's names to be more representative
- Restructures the process for more modularity
- Receives the probability map of inference outputs
- Applies non-maximal suppression
- Adds level and pixel_pacing to the input
- Removes redundant calculations
- Rewrites compute_evaluation_mask, now is compute_multi_instance_mask
- Rewrites compute_itc_list, now is compute_isolated_tumor_cells

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/pathology/datasets.py [hunk 1]
```diff
@@ -160,3 +166,132 @@ def __init__(
...
+            "image": image,
+            "mask_shape": mask.shape,
+            "mask_locations": mask_locations.astype(int).tolist(),
+            "image_locations": image_locations.astype(int).tolist(),
+            "level": level,
+        }
+
+    def _calculate_mask_level(self, image, mask) -> Tuple[int, int]:
+        """Calculate level of the mask and its ratio with respect to the whole slide image"""
+        dim_y_img, dim_x_img, _ = image.shape
```
Comment (yiheng-wang-nv) [inline monai/apps/pathology/datasets.py:250]: Hi @behxyz , I think shape length is needed here, or please add some doc strings to disclose the spatial shapes for input images and masks. What do you think?

Comment (yiheng-wang-nv) [reply monai/apps/pathology/datasets.py:250]: missing types for this and the following functions.

File: monai/utils/prob_nms.py [hunk 2]
```diff
@@ -22,8 +22,9 @@ class ProbNMS:
         prob_threshold: the probability threshold, the function will stop searching if
             the highest probability is no larger than the threshold. The value should be
             no less than 0.0. Defaults to 0.5.
-        box_size: determines the sizes of the removing area of the selected coordinates for
-            each dimensions. Defaults to 48.
+        box_size: the box size (in pixel) to be removed around the the pixel with the maximum probability.
+            It can an integer that defines the size of a square or cube,
```
Comment (yiheng-wang-nv) [inline monai/utils/prob_nms.py:26]: It can be an integer

File: monai/apps/pathology/evaluators.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        self.itc_diameter = itc_diameter
+        self.eval_thresholds = eval_thresholds
+        self.image_reader = WSIReader(image_reader_name)
+        self.nms = PathologyProbNMS(
+            sigma=0.0,
+            prob_threshold=0.5,
+            box_size=48,
+        )
+
+    def _load_data(self, file_path):
```
Comment (yiheng-wang-nv) [inline monai/apps/pathology/evaluators.py:73]: missing types for this and the following functions

File: monai/apps/pathology/datasets.py (hunk 1 above)
Comment (behxyz) [reply monai/apps/pathology/datasets.py:250]: Sure, thanks

Comment (behxyz) [reply monai/apps/pathology/datasets.py]: I addressed it in #1869

File: monai/apps/pathology/evaluators.py (hunk 3 above)
Comment (behxyz) [reply monai/apps/pathology/evaluators.py]: Done!

File: monai/utils/prob_nms.py (hunk 2 above)
Comment (behxyz) [reply monai/utils/prob_nms.py]: Done!

**Comment by behxyz** [comment]
@yiheng-wang-nv, is there any other issue that I can address before merging?

File: monai/apps/pathology/metrics.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        self.itc_diameter = itc_diameter
+        self.eval_thresholds = eval_thresholds
+        self.image_reader = WSIReader(image_reader_name)
+        self.nms = PathologyProbNMS(
+            sigma=0.0,
+            prob_threshold=0.5,
+            box_size=48,
+        )
+
+    def _load_data(self, file_path: str) -> List[Dict]:
```
Comment (wyli) [inline monai/apps/pathology/metrics.py:73]: could you remove `_load_data`, this data list should be read before this metric computation, to be consistent with the other metric modules

File: monai/apps/pathology/metrics.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    def compute_multi_instance_mask(self, mask: np.ndarray, threshold: float):
+        """
+        This method computes the segmentation mask according to the binary tumor mask.
+
+        Args:
+            mask: the binary mask array
+            threshold: the threashold to fill holes
+        """
+        # make sure it is between 0 and 1
+        assert 0 <= mask.max() <= 1, "The input mask should be a binary mask!"
```
Comment (wyli) [inline monai/apps/pathology/metrics.py:186]: remove the assert https://deepsource.io/gh/Project-MONAI/MONAI/run/2804f1da-5ea2-4956-940e-a217daca55b5/python/BAN-B101

File: monai/apps/pathology/metrics.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        # compute FROC score give specific evaluation threshold
+        froc_score = compute_froc_score(
+            fps_per_image=fps_per_image,
+            total_sensitivity=total_sensitivity,
+            eval_thresholds=self.eval_thresholds,
+        )
+
+        return froc_score
+
+    def compute_multi_instance_mask(self, mask: np.ndarray, threshold: float):
```
Comment (wyli) [inline monai/apps/pathology/metrics.py:177]: this should be a generic postprocessing utility as `self` is not used.

File: monai/apps/pathology/metrics.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        neg = 255 - mask * 255
+        distance = ndimage.morphology.distance_transform_edt(neg)
+        binary = distance < threshold
+
+        filled_image = ndimage.morphology.binary_fill_holes(binary)
+        multi_instance_mask = measure.label(filled_image, connectivity=2)
+
+        return multi_instance_mask
+
+    def compute_isolated_tumor_cells(self, tumor_mask: np.ndarray, threshold: float) -> List[int]:
```
Comment (wyli) [inline monai/apps/pathology/metrics.py:196]: this should be a generic postprocessing utility as `self` is not used.

File: monai/apps/pathology/metrics.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+measure, _ = optional_import("skimage.measure")
+ndimage, _ = optional_import("scipy.ndimage")
+
+
+class LesionFROC:
+    """
+    Evaluate with Free Response Operating Characteristic (FROC) score.
+
+    Args:
+        data: either the list of dictionaries containg probability maps (inference result) and
```
Comment (wyli) [inline monai/apps/pathology/metrics.py:31]: typo "containg"

File: monai/apps/pathology/metrics.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        return (
+            np.array(total_fp_probs),
+            np.array(total_tp_probs),
+            total_num_targets,
+            num_images,
+        )
+
+    def evaluate(self):
+        """
+        Evalaute the detection performance of a model based on the model probability map output,
```
Comment (wyli) [inline monai/apps/pathology/metrics.py:154]: typo Evalaute

File: monai/apps/pathology/metrics.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+        return froc_score
+
+    def compute_multi_instance_mask(self, mask: np.ndarray, threshold: float):
+        """
+        This method computes the segmentation mask according to the binary tumor mask.
+
+        Args:
+            mask: the binary mask array
+            threshold: the threashold to fill holes
```
Comment (wyli) [inline monai/apps/pathology/metrics.py:183]: typo threashold

File: tests/test_lesion_froc.py [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+_, has_sp = optional_import("scipy.ndimage")
+PILImage, has_pil = optional_import("PIL.Image")
+
+
+def save_as_tif(filename, array):
+    array = array[::-1, ...]  # Upside-down
+    img = PILImage.fromarray(array)
+    if not filename.endswith(".tif"):
+        filename += ".tif"
+    img.save(os.path.join("tests", "testing_data", filename))
```
Comment (wyli) [inline tests/test_lesion_froc.py:22]: please  update `.gitignore` for these new files

**Review by wyli — COMMENTED** [review]
thanks I put some comments inline

File: monai/apps/pathology/metrics.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        else:
+            self.data = data
+        self.grow_distance = grow_distance
+        self.itc_diameter = itc_diameter
+        self.eval_thresholds = eval_thresholds
+        self.image_reader = WSIReader(image_reader_name)
+        self.nms = PathologyProbNMS(
+            sigma=0.0,
+            prob_threshold=0.5,
+            box_size=48,
```
[the remark this answers is not in the document]
Comment (behxyz) [reply monai/apps/pathology/metrics.py:70]: I agree!

File: tests/test_lesion_froc.py (hunk 11 above)
Comment (behxyz) [reply tests/test_lesion_froc.py:22]: I prepended the images with `temp_` and add `tests/testing_data/temp_*` to `.gitignore`

File: monai/apps/pathology/metrics.py (hunk 4 above)
Comment (wyli) [reply monai/apps/pathology/metrics.py:78]: this is to be consistent across the project. the data working group is proposing prototypes for data representation, before we have a concrete design we shouldn't assume a JSON encoded file here

**Review by wyli — APPROVED** [review]
thanks!

**Comment by kevinkwshin** [comment]
can this module apply to 3d FROC?

**Comment by behxyz** [comment]
> can this module apply to 3d FROC?

Hi @kevinkwshin, thank for your question. This FROC is particular to pathology applications so it only supports 2D FROC.

**Pull request merged**